### PR TITLE
[KernelGen][Nvidia] Add _native_batch_norm_legit_functional operator with Triton kernel

### PR DESCRIPTION
**Pull request by XDYuanzhuLee**
## Summary
Adds a Triton kernel for `_native_batch_norm_legit_functional`. Functional version of batch normalization that returns the saved mean and inverse standard deviation, along with updated running estimates.

## Testing
- Validated against reference on device via `to_reference(inp, True)`
- Tested on: Nvidia, Tianshu, Muxi, Ascend, Hygon

## Performance
Test command: `pytest benchmark/test_native_batch_norm_legit_functional.py --level core` (NVIDIA H20)

### native_batch_norm_legit_functional

| dtype | Size | Torch Latency (ms) | Gems Latency (ms) | Speedup |
|-------|------|--------------------|-------------------|---------|
| float16 | [4, 16, 64, 4] | 0.016960 | 0.030112 | 0.563 |
| float16 | [16, 16, 8, 48] | 0.023456 | 0.031072 | 0.755 |
| float16 | [16, 16, 8, 88] | 0.030400 | 0.032768 | 0.928 |
| float16 | [16, 16, 128] | 0.017696 | 0.026720 | 0.662 |
| float16 | [20, 6, 65536] | 1.076128 | 0.343328 | 3.134 |
| float32 | [4, 16, 64, 4] | 0.017024 | 0.031968 | 0.533 |
| float32 | [16, 16, 8, 48] | 0.023488 | 0.032416 | 0.725 |
| float32 | [16, 16, 8, 88] | 0.030112 | 0.032128 | 0.937 |
| float32 | [16, 16, 128] | 0.017344 | 0.027904 | 0.622 |
| float32 | [20, 6, 65536] | 1.091840 | 0.534976 | 2.041 |
| bfloat16 | [4, 16, 64, 4] | 0.017248 | 0.031648 | 0.545 |
| bfloat16 | [16, 16, 8, 48] | 0.023392 | 0.031424 | 0.744 |
| bfloat16 | [16, 16, 8, 88] | 0.030304 | 0.033856 | 0.895 |
| bfloat16 | [16, 16, 128] | 0.017760 | 0.026528 | 0.669 |
| bfloat16 | [20, 6, 65536] | 1.075616 | 0.355488 | 3.026 |

| Operator | Arithmetic Mean Speedup |
|----------|------------------------|
| native_batch_norm_legit_functional | **1.119** |

## Multi-backend Testing
| Backend | Accuracy Test | Speedup (mean) | Notes |
|---|---|---|---|
| Nvidia (H20) | PASS (15 cases) | 1.119 | Primary |
| Tianshu | FAIL | 1.002 | Skipped: environmental/device mismatch |
| Muxi | PASS | 1.001 | — |
| Ascend | FAIL | — | Skipped: environmental/device mismatch |
| Hygon | PASS | 1.004 | — |

## Files Changed
- `src/flag_gems/ops/_native_batch_norm_legit_functional.py`: Triton kernel implementation
- `tests/test_native_batch_norm_legit_functional.py`: Accuracy test
- `benchmark/test_native_batch_norm_legit_functional.py`: Performance benchmark
- `src/flag_gems/ops/__init__.py`: Register import and `__all__`
- `src/flag_gems/__init__.py`: Register to `_FULL_CONFIG`
- `conf/operators.yaml`: Add operator entry (kind: NeuralNetwork, stage: alpha 5.4)

### CODE REVIEW

**Review by Dongxu-H — COMMENTED** [review]
## 💡 审查建议

建议:
- ⚠️ 建议在flag_gems/__init__.py注册

**Review by Dongxu-H — COMMENTED** [review]
## 💡 审查建议

建议:
- ⚠️ 建议在flag_gems/__init__.py注册

**Review by Dongxu-H — COMMENTED** [review]
## 💡 审查建议

建议:
- ⚠️ 建议在flag_gems/__init__.py注册

**Review by Dongxu-H — CHANGES_REQUESTED** [review]
## Code Review: PR with Merge Conflicts

### 📋 Checklist Review

| # | Item | Status | Notes |
|---|------|--------|-------|
| 1 | CI Status | ⚠️ CHECK | May be blocked by conflicts |
| 2 | Merge Conflicts | ❌ BLOCKING | Has merge conflicts |
| 3 | Code Review | ⏸️ DEFERRED | Until conflicts resolved |

### ⚠️ Blocking Issue

This PR has **merge conflicts** that must be resolved before it can be merged:

```
Merge State: CONFLICTING
```

### 📋 Steps to Resolve

1. **Update your branch with the latest base branch**:
   ```bash
   git fetch origin
   git checkout <your-branch>
   git merge origin/main  # or the target branch
   ```

2. **Resolve conflicts manually**:
   - Open conflicted files
   - Look for conflict markers (`<<<<<<<`, `=======`, `>>>>>>>`)
   - Choose the correct code sections
   - Remove conflict markers

3. **Commit the resolved changes**:
   ```bash
   git add .
   git commit -m "Resolve merge conflicts"
   git push
   ```

### 📊 Summary

| Category | Status |
|----------|--------|
| Merge Conflicts | ❌ **BLOCKING** |
| CI Status | ⏸️ Pending conflict resolution |
| Code Review | ⏸️ Deferred |

**Overall**: This PR cannot be merged in its current state. Please resolve the merge conflicts first, then request a new review.

**Verdict**: 🔴 **Merge Conflicts - Please Resolve Before Review**

**Review by Dongxu-H — APPROVED** [review]

**Review by Dongxu-H — CHANGES_REQUESTED** [review]
## 代码审查结果

### 🟡 建议改进

**1. [src/flag_gems/ops/_native_batch_norm_legit_functional.py] 未使用的 backward kernel**

该文件定义了 `batch_norm_backward_kernel`，但在 `_native_batch_norm_legit_functional` 函数中从未调用。这是死代码（dead code）。

```python
# 第 159-258 行定义了 backward kernel
@libentry()
@triton.autotune(...)
@triton.heuristics(...)
@triton.jit
def batch_norm_backward_kernel(...):
    ...
```

建议：
- 如果 backward kernel 是为后续 PR 准备的，请在此 PR 中移除，避免引入未测试的代码
- 如果需要在此 PR 中使用 backward，请添加相应的测试

**2. [tests/test_native_batch_norm_legit_functional.py] 缺少 backward 测试**

测试文件只测试了 forward 路径，如果 backward kernel 保留，需要添加相应的测试。

---

### 🟢 通过的检查项

| 检查项 | 状态 |
|--------|------|
| CI 状态 | ✅ PASS |
| 合并冲突 | ✅ 无冲突 |
| 配置排序 (operators.yaml) | ✅ 按字母顺序 |
| 配置排序 (__init__.py) | ✅ 按字母顺序 |
| Logger 命名 (主文件) | ✅ `logging.getLogger(__name__)` |
| Logger 命名 (iluvatar 后端) | ✅ 与现有文件一致 |
| Kernel 注册 | ✅ 正确注册 |
| Kernel 导入 | ✅ 正确导入 |
| 测试覆盖 (forward) | ✅ 覆盖多种 shape 和 dtype |
| 厂商后端文件 | ✅ iluvatar 有特殊实现 (`tl_extra_shim.rsqrt` vs `tl.math.rsqrt`) |
| 文档说明 | ✅ operators.yaml 有描述 |
| Benchmark 函数命名 | ✅ 可接受 |

---

### 📋 总结

PR 整体质量良好，主要问题是定义了但未使用的 backward kernel。建议移除该 kernel 或在后续 PR 中单独添加 backward 支持。

**Review by XDYuanzhuLee — COMMENTED** [review]
Removed unused `batch_norm_backward_kernel` (dead code, defined but never called by `_native_batch_norm_legit_functional`). The operator only supports forward path; backward tests are therefore not applicable since no backward kernel exists. The existing test file covers forward with multiple shapes, dtypes, and affine/no-affine configurations.

**Review by Dongxu-H — DISMISSED** [review]
## Review

✅ Previous review issues addressed

**Fixes applied**:
- Merge conflicts resolved
- Unused backward kernel removed (commit 7806ca6)
- Operator registered in __init__.py
- Iluvatar backend specialization added

**Implementation**: ✅ Proper batch norm kernel with training/inference modes, running stats tracking

**Performance**: 1.119x mean speedup (good)
- Best: 3.134x (float16, [20, 6, 65536])
- Worst: 0.533x (float32, [4, 16, 64, 4])

**CI**: ✅ All checks passing (python-op, backend-tests, code-style)

**Multi-backend**:
- Nvidia: ✅ PASS (15 cases)
- Muxi, Hygon: ✅ PASS
- Tianshu, Ascend: ❌ FAIL (environmental/device mismatch, not code issue)

**Test coverage**: ✅ Multiple shapes (3D, 4D), all float dtypes, training/inference modes

**Code quality**: ✅ Proper structure, attribution, kernel copied from attorch and modified

**Verdict**: Approve - all review feedback addressed, CI passes, implementation correct.

File: tests/test_native_batch_norm_legit_functional.py [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+            running_var,
+            True,
+            0.1,
+            eps,
+        )
+
+    utils.gems_assert_close(res_out, ref_out, dtype)
+    utils.gems_assert_close(res_save_mean, ref_save_mean, dtype)
+    utils.gems_assert_close(res_running_mean_out, ref_running_mean_out, dtype)
+    utils.gems_assert_close(res_running_var_out, ref_running_var_out, dtype)
```
Comment (Caeruleann) [inline tests/test_native_batch_norm_legit_functional.py:85]: The test doesn't assert save_var — please add gems_assert_close(res_save_var, ref_save_var)

Comment (kkkwb) [reply tests/test_native_batch_norm_legit_functional.py:85]: Thanks. Added the save_var assertion. Commit 91f8c419ae84ae5d93fefad824d7d211689cf927.

**Review by Caeruleann — DISMISSED** [review]

**Review by huangyiqun — APPROVED** [review]
LGTM.
